i2c/mpl3115a2_i2c/mpl3115a2_i2c.c

```diff
@@ -0,0 +1,203 @@
+        // otherwise, we would read F_STATUS to confirm it was an overflow
+        printf("FIFO overflow!\n");
+        // drain the fifo
+        mpl3115a2_read_fifo(fifo_data);
+        // read status register to clear interrupt bit
+        mpl3115a2_read_reg(MPL3115A2_F_STATUS);
+        has_new_data = true;
+    }
+}
+#endif
+
+void mpl3115a2_convert_fifo_batch(uint8_t start, volatile uint8_t buf[], struct mpl3115a2_data_t* data) {
+    // convert a batch of fifo data into temperature and altitude data
+
+    // 3 altitude registers: MSB (8 bits), CSB (8 bits) and LSB (4 bits, starting from MSB)
+    // first two are integer bits (2's complement) and LSB is fractional bits -> makes 20 bit signed integer
+    int32_t h = (int32_t)((uint32_t)buf[start] << 24 | buf[start + 1] << 16 | buf[start + 2] << 8);
+    data->altitude = h * 1.f / 65536;
+
+    // 2 temperature registers: MSB (8 bits) and LSB (4 bits, starting from MSB)
+    // first 8 are integer bits with sign and LSB is fractional bits -> 12 bit signed integer
+    int16_t t = (int16_t)(((uint16_t)buf[start + 3]) << 8 | buf[start + 4]);
+    data->temperature = t * 1.f / 256;
+}
+
+int main() {
+    stdio_init_all();
+#if !defined(i2c_default) || !defined(PICO_DEFAULT_I2C_SDA_PIN) || !defined(PICO_DEFAULT_I2C_SCL_PIN)
+    #warning i2c / mpl3115a2_i2c example requires a board with I2C pins
+        puts("Default I2C pins were not defined");
+#else
+    printf("Hello, MPL3115A2. Waiting for something to interrupt me!...\n");
+
+    // use default I2C0 at 400kHz, I2C is active low
+    i2c_init(i2c_default, 400 * 1000);
+    gpio_set_function(PICO_DEFAULT_I2C_SDA_PIN, GPIO_FUNC_I2C);
+    gpio_set_function(PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C);
+    gpio_pull_up(PICO_DEFAULT_I2C_SDA_PIN);
+    gpio_pull_up(PICO_DEFAULT_I2C_SCL_PIN);
+
+    gpio_init(INT1_PIN);
+    gpio_pull_up(INT1_PIN); // pull it up even more!
+
+    // add program information for picotool
+    bi_decl(bi_program_name("Example in the pico-examples library for the MPL3115A2 altimeter"));
+    bi_decl(bi_1pin_with_name(16, "Interrupt pin 1"));
+    bi_decl(bi_2pins_with_func(PICO_DEFAULT_I2C_SDA_PIN, PICO_DEFAULT_I2C_SCL_PIN, GPIO_FUNC_I2C));
+
+    mpl3115a2_init();
+
+    gpio_set_irq_enabled_with_callback(INT1_PIN, GPIO_IRQ_LEVEL_LOW, true, &gpio_callback);
+
+    while (1) {
+        // as interrupt data comes in, let's print the 32 sample average
+        if (has_new_data) {
+            float tsum = 0, hsum = 0;
+            struct mpl3115a2_data_t data;
+            for (int i = 0; i < MPL3115A2_FIFO_SIZE; i++) {
+                mpl3115a2_convert_fifo_batch(i * MPL3115A2_DATA_BATCH_SIZE, fifo_data, &data);
+                tsum += data.temperature;
+                hsum += data.altitude;
+            }
+            printf("%d sample average -> t: %.4f C, h: %.4f m\n", MPL3115A2_FIFO_SIZE, tsum / MPL3115A2_FIFO_SIZE, hsum / MPL3115A2_FIFO_SIZE);
+            has_new_data = false;
```

Comment:
There's no mutex protection on the has_new_data variable and it is set in the interrupt routine and tested here. However, its likely OK in this simple case.